OpenRA.Mods.Common/Traits/Player/EnemyWatcher.cs

```diff
@@ -0,0 +1,117 @@
+		public readonly int ScanInterval = 25;
+
+		[Desc("Minimal interval in ticks between notifications.")]
+		public readonly int NotificationInterval = 200;
+
+		public object Create(ActorInitializer init) { return new EnemyWatcher(init.Self, this); }
+	}
+
+	class EnemyWatcher : ITick
+	{
+		readonly EnemyWatcherInfo info;
+		readonly Lazy<RadarPings> radarPings;
+
+		bool announcedAny;
+		int rescanInterval;
+		int ticksBeforeNextNotification;
+		HashSet<uint> lastKnownActorIds;
+		HashSet<uint> visibleActorIds;
+		HashSet<string> playedNotifications;
+
+		public EnemyWatcher(Actor self, EnemyWatcherInfo info)
+		{
+			lastKnownActorIds = new HashSet<uint>();
+			this.info = info;
+			rescanInterval = info.ScanInterval;
+			ticksBeforeNextNotification = info.NotificationInterval;
+			radarPings = Exts.Lazy(() => self.World.WorldActor.Trait<RadarPings>());
+		}
+
+		public void Tick(Actor self)
+		{
+			// TODO: Make the AI handle such notifications and remove Owner.IsBot from this check
+			// Disable notifications for AI and neutral players (creeps) and for spectators
+			if (self.Owner.Shroud.Disabled || self.Owner.IsBot || !self.Owner.Playable || self.Owner.PlayerReference.Spectating)
+				return;
+
+			rescanInterval--;
+			ticksBeforeNextNotification--;
+
+			if (rescanInterval > 0 || ticksBeforeNextNotification > 0)
+				return;
+
+			rescanInterval = info.ScanInterval;
+
+			announcedAny = false;
+			visibleActorIds = new HashSet<uint>();
+			playedNotifications = new HashSet<string>();
+
+			foreach (var actor in self.World.ActorsWithTrait<AnnounceOnSeen>())
+			{
+				// We only care about enemy actors (creeps should be enemies)
+				if ((actor.Actor.EffectiveOwner != null && self.Owner.Stances[actor.Actor.EffectiveOwner.Owner] != Stance.Enemy)
+				 || self.Owner.Stances[actor.Actor.Owner] != Stance.Enemy)
+					continue;
+
+				// The actor is not currently visible
+				if (!self.Owner.Shroud.IsVisible(actor.Actor))
+					continue;
+
+				visibleActorIds.Add(actor.Actor.ActorID);
+
+				// We already know about this actor
+				if (lastKnownActorIds.Contains(actor.Actor.ActorID))
+					continue;
+
+				// We have already played this type of notification
+				if (playedNotifications.Contains(actor.Trait.Info.Notification))
+					continue;
+
+				Announce(self, actor);
+			}
+
+			if (announcedAny)
+				ticksBeforeNextNotification = info.NotificationInterval;
+
+			lastKnownActorIds = visibleActorIds;
+		}
+
+		void Announce(Actor self, TraitPair<AnnounceOnSeen> announce)
+		{
+			// Audio notification
+			if (self.World.LocalPlayer != null)
+				Sound.PlayNotification(self.World.Map.Rules, self.World.LocalPlayer, "Speech", announce.Trait.Info.Notification, self.Owner.Country.Race);
+
+			// Radar notificaion
+			if (announce.Trait.Info.PingRadar && radarPings.Value != null)
+				radarPings.Value.Add(() => true, announce.Actor.CenterPosition, Color.Red, 50);
+
+			Game.Debug(announce.Trait.Info.Notification);
```

Comment:
Ah, I added this because I was testing on a PC where for some reason OpenRA's sound doesn't work, so I needed another way to keep track, other than breakpoints.
I think it would be better to remove it.